Source/WebCore/html/ImageBitmapBacking.cpp

```diff
@@ -72,8 +72,14 @@ void ImageBitmapBacking::disconnect()
     // FIXME: Rather than storing both the ImageBuffer and the
     // SerializedImageBuffer here, with only one valid at a time,
     // we should have a separate object for the serialized state
-    if (m_bitmapData)
-        m_serializedBitmap = ImageBuffer::sinkIntoSerializedImageBuffer(WTFMove(m_bitmapData));
+    if (m_bitmapData) {
+        if (m_bitmapData->hasOneRef())
+            m_serializedBitmap = ImageBuffer::sinkIntoSerializedImageBuffer(WTFMove(m_bitmapData));
+        else {
+            m_serializedBitmap = ImageBuffer::sinkIntoSerializedImageBuffer(m_bitmapData->clone());
+            m_bitmapData = nullptr;
+        }
+    }
```

Comment:
Is this change related to this patch? If it is not please do it in a separate patch.

Also I think it is better to move this optimization to `ImageBuffer::sinkIntoSerializedImageBuffer()`:

```
std::unique_ptr<SerializedImageBuffer> ImageBuffer::sinkIntoSerializedImageBuffer(RefPtr<ImageBuffer>&& image)
{
    RefPtr<ImageBuffer> move;
    if(image->hasOneRef())
        move = WTFMove(image);
    else
        move = image->clone();
      ... 
}
```

Reply:
It is! I tried to explain in the commit message, but now that a `Pattern` object can hold a reference to the `ImageBuffer` within an `ImageBitmap`, we can no longer guarantee it has one ref (and we assert at runtime).

My mental model for 'sink into' was that in consumes the object, which is why it takes an `&&`. That's not super well defined for a ref counted object obviously, which is why it has a runtime assert instead.

Having the `ImageBuffer` implementation handle the copy is certainly easier, but maybe it'd be worth changing the function name too? Or I need to change my mental model of what 'sink into' means :)